crypto/arm_arch.h

```diff
@@ -49,6 +49,10 @@
 #   else
 #    error "unsupported ARM architecture"
 #   endif
+#  else
+#   if  defined(__ARM_ARCH)
```

Comment:
There's an extra space between `if` and `defined(__ARM_ARCH)`, and it might be better if this checked `__ARM_ARCH__` too

```
#   if defined(__ARM_ARCH) && !defined(__ARM_ARCH__)
```

Reply:
Err, sorry, upon closer inspection this is already under `!defined(__ARM_ARCH__)` at line 13, so that bit's not needed.

However, it would match the code above better using `elif`, and I also wonder if we should check `_M_ARM64` with something like

```
@@ -49,6 +49,10 @@
 #   else
 #    error "unsupported ARM architecture"
 #   endif
+#  elif defined(_M_ARM64) || defined(_M_ARM64EC)
+#    define __ARM_ARCH__ 8
+#  elif defined(__ARM_ARCH)
+#    define __ARM_ARCH__ __ARM_ARCH
 #  endif
 # endif

```

Reply:
White spacing has been fixed.
__ARM_ARCH should specify the version of the target architecture, so I would prefer to keep __ARM_ARCH usage in this PR.
_M_ARM64EC check could be a good idea which deserves an additional PR, testing and benchmarking.

Reply:
Agree that testing for `_M_ARM64(EC)` in this file makes sense, for compiling with MSVC

Reply:
Right, in my head this definition is just about correctly identifying the target architecture at compile time, but on closer inspection it is about enabling use of OpenSSL's optimized perlasm.

I think more work needs to be done to enable perlasm use with:
a) MSVC assembler on Arm64 ([armasm64](https://learn.microsoft.com/en-us/cpp/assembler/arm/arm-assembler-command-line-reference?view=msvc-170))
b) [Arm64EC ABI](https://learn.microsoft.com/en-us/windows/arm/arm64ec-abi)

So this PR is just helpful for enabling optimizations for:
a) native Arm64 binaries on Windows (as Arm64 ABI on Windows is mostly AArch64 EABI)
b) using clang-cl (with GNU-like assembler)

Someone needs to check closely that using clang-cl + perlasm generates appropriate metadata (.pdata/.xdata records) for [Windows' structured exception handling](https://learn.microsoft.com/en-us/cpp/build/arm64-exception-handling?view=msvc-170).

Reply:
Resolving this conversion as it was discussed earlier regarding Arm64EC support.
https://github.com/openssl/openssl/pull/21673#discussion_r1333325839